pkg/apis/options/load.go

```diff
@@ -0,0 +1,133 @@
+package options
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	"github.com/mitchellh/mapstructure"
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+)
+
+// Load reads in the config file at the path given, then merges in environment
+// variables (prefixed with `OAUTH2_PROXY`) and finally merges in flags from the flagSet.
+// If a config value is unset and the flag has a non-zero value default, this default will be used.
+// Eg. A field defined:
+//    FooBar `cfg:"foo_bar" flag:"foo-bar"`
+// Can be set in the config file as `foo_bar="baz"`, in the environment as `OAUTH2_PROXY_FOO_BAR=baz`,
+// or via the command line flag `--foo-bar=baz`.
+func Load(configFileName string, flagSet *pflag.FlagSet, into interface{}) error {
+	v := viper.New()
+	v.SetConfigFile(configFileName)
+	v.SetConfigType("toml") // Config is in toml format
+	v.SetEnvPrefix("OAUTH2_PROXY")
+	v.AutomaticEnv()
+	v.SetTypeByDefaultValue(true)
+
+	if configFileName != "" {
+		err := v.ReadInConfig()
+		if err != nil {
+			return fmt.Errorf("unable to load config file: %w", err)
+		}
+	}
+
+	err := registerFlags(v, "", flagSet, into)
+	if err != nil {
+		// This should only happen if there is a programming error
+		return fmt.Errorf("unable to register flags: %w", err)
+	}
+
+	err = v.Unmarshal(into, decodeFromCfgTag)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling config: %w", err)
+	}
+
+	return nil
+}
+
+// registerFlags uses `cfg` and `flag` tags to associate flags in the flagSet
+// to the fields in the options interface provided.
+// Each exported field in the options must have a `cfg` tag otherwise an error will occur.
+// - For fields, set `cfg` and `flag` so that `flag` is the name of the flag associated to this config option
+// - For exported fields that are not user facing, set the `cfg` to `,internal`
+// - For structs containing user facing fields, set the `cfg` to `,squash`
+func registerFlags(v *viper.Viper, prefix string, flagSet *pflag.FlagSet, options interface{}) error {
+	val := reflect.ValueOf(options)
+	var typ reflect.Type
+	if val.Kind() == reflect.Ptr {
+		typ = val.Elem().Type()
+	} else {
+		typ = val.Type()
+	}
+
+	for i := 0; i < typ.NumField(); i++ {
+		// pull out the struct tags:
+		//    flag - the name of the command line flag
+		//    cfg - the name of the config file option
+		//    env - the name of the environment variable
```

Comment:
Looks like you are using automatic env instead

Reply:
Good catch with fix